pkg/util/discardedseries/tracker.go

```diff
@@ -0,0 +1,102 @@
+package discardedseries
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var trackedLabels = []string{
+	"sample_out_of_bounds",
+	"sample_out_of_order",
+	"sample_too_old",
+	"new_value_for_timestamp",
+	"per_user_series_limit",
+	"per_labelset_series_limit",
+	"per_metric_series_limit",
+}
+var vendMetricsInterval = 30 * time.Second
+
+type SeriesCounter struct {
+	*sync.RWMutex
+	seriesCountMap map[uint64]struct{}
+}
+
+type UserCounter struct {
+	*sync.RWMutex
+	userSeriesMap map[string]*SeriesCounter
+}
+
+type DiscardedSeriesTracker struct {
+	labelUserMap         map[string]*UserCounter
+	discardedSeriesGauge *prometheus.GaugeVec
+}
+
+func NewDiscardedSeriesTracker(discardedSeriesGauge *prometheus.GaugeVec) *DiscardedSeriesTracker {
+	labelUserMap := make(map[string]*UserCounter)
+	for _, label := range trackedLabels {
+		labelUserMap[label] = &UserCounter{
+			RWMutex:       &sync.RWMutex{},
+			userSeriesMap: make(map[string]*SeriesCounter),
+		}
+	}
+	return &DiscardedSeriesTracker{labelUserMap: labelUserMap, discardedSeriesGauge: discardedSeriesGauge}
+}
+
+func (t *DiscardedSeriesTracker) Track(label string, user string, series uint64) {
+	if userCounter, ok := t.labelUserMap[label]; ok {
+		seriesCounter, ok := userCounter.userSeriesMap[user]
+		if !ok {
+			seriesCounter = &SeriesCounter{
+				RWMutex:        &sync.RWMutex{},
+				seriesCountMap: make(map[uint64]struct{}),
+			}
+			userCounter.Lock()
+			userCounter.userSeriesMap[user] = seriesCounter
+			userCounter.Unlock()
+		}
+
+		if _, ok := seriesCounter.seriesCountMap[series]; !ok {
+			seriesCounter.Lock()
+			seriesCounter.seriesCountMap[series] = struct{}{}
+			seriesCounter.Unlock()
+		}
+	}
+}
+
+func (t *DiscardedSeriesTracker) UpdateMetrics() {
+	for label, userCounter := range t.labelUserMap {
+		userCounter.Lock()
```

Comment:
We maybe want to consider changing it to be UserLabel instead of LabelUser.
When emitting this metric for a Label, we would block global requests from every user.
It seems this would have faster lock if we have UserLabel.